### PR TITLE
Enable dispatching linalg.generic ops into workgroups and set tiling paramters for LLVM backend

### LINKED ISSUE #5045: Use tile and distribute for all compute operations
The current dispatch region creation uses the tile + distribute for only named ops. For elemntwise operations represented as `linalg.generic` and `linalg.indexed_generic` operations they are not tiled during dispatch region creation. Instead they are left as is and
- Lowered to loops on CPU side to execute sequentially
- On GPU side the patterns that lower an untiled operations to scalar code by distributing them to invocations using `GlobalInvocationId` kick in to distribute one iteration per invocation.

Both of this is not satisfactory. We need to parallelize the work on CPU side, and we also should have such low computation per thread on the GPU side. Instead it might be worth using tile + distribute on these operations as well to do some coarse-grain partitioning of the work. On the GPU side we can then try to use vectorization to get vector loads/stores (which also can be done on the CPU side).

Points to consider
- On GPU side we might map this directly to invocations (using global invocation id) so that we dont add unnecessary loop overhead (by default the tiles created at the flow level are mapped to workgroups)

### PR DESCRIPTION
**Pull request by asaadaldien**
Fix #5045

### CODE REVIEW

File: iree/compiler/Conversion/LinalgToLLVM/KernelDispatch.cpp [hunk 1]
```diff
@@ -129,35 +151,39 @@ DEFINE_TILE_SIZE_FN(TilingLevel::Level2Tiles)
...
 #undef DEFINE_TILE_SIZE_FN
 
+bool isDispatchOp(Operation *op) {
+  if (auto contractionOp = dyn_cast<linalg::ContractionOpInterface>(op)) {
+    if (contractionOp.isRowMajorMatmul() ||
+        contractionOp.isRowMajorBatchMatmul()) {
+      return true;
+    }
+  }
+  if (isa<linalg::GenericOp>(op)) return true;
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToLLVM/KernelDispatch.cpp:161]: What happens with indexed_generic ? We can stage this. The GPU side should be able to handle `IndexedGeneric` too, its just not plumbed through this path on the dispatch region creation step.

File: iree/compiler/Dialect/Flow/Transforms/test/dispatch_linalg_on_tensors.mlir [hunk 2]
```diff
@@ -582,16 +584,16 @@ func @inline_dag_1(
 //   CHECK-NOT:   linalg.
 //   CHECK-NOT:   subtensor
 //       CHECK:   flow.dispatch.workgroups
-//  CHECK-NEXT:     %[[ARG4:[a-zA-Z0-9_]+]]: !flow.dispatch.tensor<readonly:i32>
-//  CHECK-SAME:     %[[ARG5:[a-zA-Z0-9_]+]]: !flow.dispatch.tensor<readonly:1x?xf32>
-//  CHECK-SAME:     %[[ARG6:[a-zA-Z0-9_]+]]: !flow.dispatch.tensor<readonly:?xf32>
+//  CHECK-NEXT:     %[[ARG4:[a-zA-Z0-9_]+]]: !flow.dispatch.tensor<readonly:1x?xf32>
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/test/dispatch_linalg_on_tensors.mlir:587]: Thats annoying. Wish there was `CHECK-SAME-DAG`

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 3]
```diff
@@ -59,13 +59,6 @@ static llvm::cl::opt<bool> clEnableOperandFusion(
         "Enable fusing operand producers during dispatch region formation"),
     llvm::cl::init(false));
 
-// TODO(#5045): Tile and distribute on CPU causes performance regressions. This
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:62]: Awesome!

**Review by MaheshRavishankar — APPROVED** [review]

File: iree/compiler/Conversion/LinalgToLLVM/KernelDispatch.cpp [hunk 4]
```diff
@@ -104,6 +101,31 @@ llvm::SmallVector<int64_t, 4> getTileSizes(Operation *op) {
...
+            workgroupTileSizes.push_back(genericOpsWorkgroupTileSize);
+          } else {
+            // Don't tile workgroup across reduction dimensions.
+            workgroupTileSizes.push_back(0);
+          }
+        }
+        return workgroupTileSizes;
+      }
+      // TODO(ataei): Set the parameters when we enable vectorization.
+      case TilingLevel::Level1Tiles:
```
[the remark this answers is not in the document]
Comment (asaadaldien) [reply iree/compiler/Conversion/LinalgToLLVM/KernelDispatch.cpp:123]: For now yes, a follow up PR will change that soon :)

File: iree/compiler/Dialect/Flow/Transforms/test/dispatch_linalg_on_tensors.mlir (hunk 2 above)
Comment (asaadaldien) [reply iree/compiler/Dialect/Flow/Transforms/test/dispatch_linalg_on_tensors.mlir:587]: yeah toke me some time to rearrange the args :/

File: iree/compiler/Conversion/LinalgToLLVM/KernelDispatch.cpp (hunk 1 above)
Comment (asaadaldien) [reply iree/compiler/Conversion/LinalgToLLVM/KernelDispatch.cpp:161]: we can do that in a follow up PR, need to enable it in `DispatchLinalg..` pass too

File: iree/compiler/Conversion/LinalgToLLVM/KernelDispatch.cpp [hunk 5]
```diff
@@ -104,6 +101,31 @@ llvm::SmallVector<int64_t, 4> getTileSizes(Operation *op) {
     }
   }
 
+  if (isa<linalg::GenericOp>(op)) {
+    auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+    switch (tilingLevel) {
+      case TilingLevel::WorkGroupTiles: {
+        llvm::SmallVector<int64_t, 4> workgroupTileSizes;
+        int iterationRank = linalgOp.iterator_types().size();
```
Comment (ThomasRaoux) [inline iree/compiler/Conversion/LinalgToLLVM/KernelDispatch.cpp:109]: Can you just do `iterationRank = getNumOuterParallelLoops(linalgOp);` this way you don't pick up the reduction dimensions? Also I'm guessing you only want to tile outer parallel loops to match what is done in flow?

**Comment by MaheshRavishankar** [comment]
@asaadaldien forgot to mention. You will have to drop these flags from the mako configurations too

File: iree/compiler/Conversion/LinalgToLLVM/KernelDispatch.cpp (hunk 5 above)
Comment (asaadaldien) [reply iree/compiler/Conversion/LinalgToLLVM/KernelDispatch.cpp:109]: Right! I want to tile outer parallel loop and also skip inner reduction ones.

**Comment by antiagainst** [comment]
Nice to see this option gone and elementwise tiling and distribution reenabled! Thanks for fixing this, Ahmed!

**Comment by hanhanW** [comment]
Interesting... this seems to make a big regression on Mobilebert for CPU, so the BuildKite tests timeout. I disable the batch mode and found that it takes 11838 ms for Mobilebert on Pixel4 CPU.

https://buildkite.com/iree/iree-android-benchmark/builds/2374

**Comment by asaadaldien** [comment]
> Interesting... this seems to make a big regression on Mobilebert for CPU, so the BuildKite tests timeout. I disable the batch mode and found that it takes 11838 ms for Mobilebert on Pixel4 CPU.
> 
> https://buildkite.com/iree/iree-android-benchmark/builds/2374

Strange! there was no regression when this was landed. looking into it now.

**Comment by asaadaldien** [comment]
mmm, I can't repro this with locally I am still getting ~ `690 ms` on main branch

**Pull request merged**